prio_processor/spark/commands.py

```diff
@@ -0,0 +1,355 @@
+import math
+from base64 import b64decode, b64encode
+from datetime import datetime
+from functools import partial
+from pathlib import Path
+from uuid import uuid4
+
+import click
+from prio_processor.prio.commands import import_keys, match_server
+from prio_processor.prio.options import (
+    data_config,
+    input_1,
+    input_2,
+    output_1,
+    output_2,
+    public_key,
+    server_config,
+)
+from prio_processor.spark import udf
+from pyspark.sql import SparkSession
+from pyspark.sql import functions as F
+
+
+def spark_session():
+    spark = SparkSession.builder.getOrCreate()
+    spark.conf.set("spark.sql.session.timeZone", "UTC")
+    spark.conf.set("spark.sql.execution.arrow.pyspark.enabled", "true")
+    return spark
+
+
+@click.group()
+def entry_point():
+    pass
+
+
+@entry_point.command()
+@data_config
+@public_key
+@output_2
+@click.option("--n-rows", type=int, help="Number of rows randomly generate.")
+@click.option(
+    "--scale", type=int, default=1, help="Factor to duplicate shares for output."
+)
+@click.option("--partition-size-mb", type=int, default=250, help="")
+def generate(
+    batch_id,
+    n_data,
+    public_key_hex_internal,
+    public_key_hex_external,
+    output_a,
+    output_b,
+    n_rows,
+    scale,
+    partition_size_mb,
+):
+    shares = (
+        spark_session()
+        .range(n_rows * n_data)
+        .select(
+            (F.col("id") % n_rows).alias("row_id"),
+            F.when(F.rand() > 0.5, 1).otherwise(0).alias("payload"),
+        )
+        .groupBy("row_id")
+        .agg(F.collect_list("payload").alias("payload"))
+        .select(
+            F.pandas_udf(
+                partial(
+                    udf.encode,
+                    batch_id,
+                    n_data,
+                    public_key_hex_internal,
+                    public_key_hex_external,
+                ),
+                returnType="a: binary, b: binary",
+            )("payload").alias("shares")
+        )
+        # repeat this data `scale` times
+        .withColumn("_repeat", F.explode(F.array_repeat(F.lit(0), scale)))
+        .drop("_repeat")
+        .withColumn("id", F.udf(lambda: str(uuid4()), returnType="string")())
+    )
+    shares.cache()
+    # we can make an estimate with just a single row, since the configuration
+    # is the same here.
+    row = shares.first()
+    num_partitions = math.ceil(
+        (len(b64encode(row.shares.a)) + len(str(uuid4())))
+        * n_rows
+        * scale
+        * 1.0
+        / 10 ** 6
+        / partition_size_mb
+    )
```

Comment:
I think splitting this up into two steps makes it easier to understand:

```suggestion
    data_size = math.ceil(
        (len(b64encode(row.shares.a)) + len(str(uuid4())))
        * n_rows
        * scale
        * 1.0
        / 10 ** 6
    )
    num_partitions = data_size / partition_size_mb
```